glyphs-reader/src/glyphdata.rs

```diff
@@ -599,9 +605,73 @@ impl GlyphData {
         Self::construct_category_via_agl(base_name)
     }
 
+    // https://github.com/googlefonts/glyphsLib/blob/c4db6b981d5/Lib/glyphsLib/glyphdata.py#L351
+    fn construct_production_name(&self, name: &str) -> Option<ProductionName> {
+        fn append_suffix(base_name: &mut String, suffix: Option<&str>) {
+            if let Some(suffix) = suffix {
+                base_name.push('.');
+                base_name.push_str(suffix);
+            }
+        }
+
+        fn is_u_name(name: &str) -> bool {
+            name.starts_with("u") && name[1..].bytes().all(|b| b.is_ascii_hexdigit())
+        }
+
+        let (base_name, suffix) = self.split_glyph_suffix(name);
+        let base_names = self
+            .split_ligature_glyph_name(base_name)
+            .unwrap_or_else(|| vec![base_name.into()]);
+        // Attempt to find a production name for each ligature component (or the whole base name).
+        // Return early if any such names have no GlyphData entry
+        // OR the entry doesn't specify a production name AND they aren't already AGLFN names...
+        let prod_names: Vec<SmolStr> = base_names
+            .into_iter()
+            .map(|name| {
+                self.query_no_synthesis(&name, None).and_then(|result| {
+                    result.production_name.map(Into::into).or_else(|| {
+                        // if no production name, return the name itself if already in AGLFN
+                        fontdrasil::agl::char_for_agl_name(name.as_ref()).map(|_| name)
+                    })
+                })
+            })
+            .collect::<Option<_>>()?;
+
+        // only (uniXXXX, uniYYYY, etc.) names with 4 hex digits can be concatenated using the
+        // more compact format uniXXXXYYYY... uXXXXX names for characters beyond BMP are joined
+        // in ligatures using the usual '_'.
+        let any_characters_outside_bmp = prod_names
+            .iter()
+            .any(|name| name.len() > 5 && is_u_name(name.as_ref()));
+        let any_uni_names = prod_names.iter().any(|name| name.starts_with("uni"));
+
+        if !any_characters_outside_bmp && any_uni_names {
+            let mut uni_names = Vec::new();
+            for part in prod_names {
+                if let Some(stripped) = part.strip_prefix("uni") {
+                    uni_names.push(stripped.to_string());
+                } else if part.len() == 5 && is_u_name(part.as_ref()) {
+                    uni_names.push(part[1..].to_string());
```

Comment:
is the to_string required? You can't capture &str's?

Reply:
well, in the following else branch we do need to create owned Strings and push to the same vec.. then later on we need to join them all into one big string so I suppose we need.. I could use a Cow<str> perhaps